### PR TITLE
Ensure total tokens in system is kept up to date

### PR DESCRIPTION
**Pull request by shamardy**
WIP for https://github.com/paritytech/polkadot/issues/303

### CODE REVIEW

**Comment by gavofyork** [comment]
@shamardy let me know when this is ready for review :)

**Comment by shamardy** [comment]
@gavofyork It can be reviewed now, just thought I might add some more tests.

File: demo/executor/src/lib.rs [hunk 1]
```diff
@@ -289,24 +289,24 @@ mod tests {
...
 
-		executor().call(&mut t, COMPACT_CODE, "execute_block", &block2().0, true).0.unwrap();
+	//	executor().call(&mut t, COMPACT_CODE, "execute_block", &block2().0, true).0.unwrap();
 
-		runtime_io::with_externalities(&mut t, || {
-			assert_eq!(Staking::voting_balance(&alice()), 30);
-			assert_eq!(Staking::voting_balance(&bob()), 78);
-		});
-	}
+	//	runtime_io::with_externalities(&mut t, || {
```
Comment (gavofyork) [inline demo/executor/src/lib.rs]: commenting should be block comments `/* ... */`, but in any case, we don't generally check in commented code.

File: substrate/runtime/staking/src/lib.rs [hunk 2]
```diff
@@ -424,7 +424,9 @@ impl<T: Trait> Module<T> {
 			Self::on_reserved_too_low(who);
 			UpdateBalanceOutcome::AccountKilled
 		} else {
+			Self::decrease_total_stake(Self::reserved_balance(who));
```
Comment (gavofyork) [inline substrate/runtime/staking/src/lib.rs]: yeah - this is one way of doing it, but it's rather inefficient. a better way might be to just check where this function can be called and ensure that they look after the increase/reduction in stake.

Comment (shamardy) [reply substrate/runtime/staking/src/lib.rs]: I thought this way will save anyone using this function in future code from updating the TotalStake value by calling the decrease or increase function. This is the case with the similar function `on_free_too_low` which is called 2 times in the code. what do you think?

Comment (shamardy) [reply substrate/runtime/staking/src/lib.rs]: I thought this way will save anyone using the `set_reserved_balance` function in future code from updating the TotalStake value by calling the increase or decrease function every time. Same with the `set_free_balance` which is similar and is more through the code.

**Comment by shamardy** [comment]
@gavofyork about the `full_native_block_import_works` , `native_big_block_import_succeeds` test functions that failed in the last commit due to the change in storage root, isn't the storage root supposed to change now due to the change in TotalStake in the storage? what should I do with those tests?

**Comment by gavofyork** [comment]
@shamardy yeah - as long as that's the only thing that changed, you can safely alter the expected root in those tests.

**Comment by gavofyork** [comment]
looks good! just needs the tests updating. (sorry about the delay with the review - lot going on currently.)

**Comment by shamardy** [comment]
[clabot:check]

File: substrate/runtime/session/src/lib.rs [hunk 3]
```diff
@@ -142,6 +142,11 @@ impl<T: Trait> Module<T> {
 	pub fn force_new_session(normal_rotation: bool) -> Result {
 		<ForcingNewSession<T>>::put(normal_rotation);
 		Ok(())
+	}	
+
+	/// Sets bad validators.
+	pub fn set_bad_validators(bad_validators: Vec<T::AccountId>) {
```
Comment (gavofyork) [inline substrate/runtime/session/src/lib.rs]: any reason for this?

Comment (shamardy) [reply substrate/runtime/session/src/lib.rs]: used it in a past commit to set bad validators for early era slashing, not needed now after last update, will remove it.

**Comment by gavofyork** [comment]
[clabot:check]

**Pull request merged**